File: python/bob/bob.py
# -*- coding: utf-8 -*-

def hey(phrase):
  if is_silence(phrase):
    return "Fine. Be that way!"
  elif is_shouting(phrase):
    return "Whoa, chill out!"
  elif is_question(phrase):
    return "Sure."
  else:
    return "Whatever."

def is_silence(phrase):
  return not phrase.strip()

def is_question(phrase):
  return phrase.strip().endswith("?")

def is_shouting(phrase):
  return has_letter(phrase) and phrase.upper() == phrase

def has_letter(phrase):
  return any([c.isalpha() for c in phrase])

File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(phrase) do
    phrase = String.strip(phrase)

    cond do
      silence?(phrase) ->
        "Fine. Be that way!"
      question?(phrase) && shout?(phrase) ->
        "Calm down, I know what I'm doing!"
      shout?(phrase) ->
        "Whoa, chill out!"
      question?(phrase) ->
        "Sure."
      true ->
        "Whatever."
    end
  end

  defp silence?(phrase), do: phrase == ""  

  defp shout?(phrase), do: has_letter?(phrase) && String.upcase(phrase) == phrase 

  defp has_letter?(phrase), do: String.match?(phrase, ~r/\p{L}+/)  

  defp question?(phrase), do: String.ends_with?(phrase, "?")     
end
